lib/Dialect/ESI/ESIPasses.cpp

```diff
@@ -245,19 +300,242 @@ LogicalResult PipelineStageLowering::matchAndRewrite(
+/// to the inputs and remove the output channel from the results. Then call the
+/// function which adapts the instances (the hard part).
+void ESIPortsPass::updateFunc(RTLExternModuleOp mod) {
+  auto *ctxt = &getContext();
+  auto funcType = mod.getType();
+
+  bool updated = false;
+
+  // Reconstruct the list of operand types, changing the type whenever an ESI
+  // port is found.
+  SmallVector<Type, 16> newArgTypes;
+  for (auto argTy : funcType.getInputs()) {
+    auto chanTy = argTy.dyn_cast<ChannelPort>();
+    if (!chanTy) {
+      newArgTypes.push_back(argTy);
+      continue;
+    }
+
+    // When we find one, construct an interface, and add the 'sink' modport to
+    // the type list.
+    auto iface = B->getOrConstructInterface(chanTy);
+    newArgTypes.push_back(iface.getModportType(B->sink));
+    updated = true;
+  }
+
+  SmallVector<MutableDictionaryAttr, 16> argAttrs;
+  mod.getAllArgAttrs(argAttrs);
+  SmallVector<MutableDictionaryAttr, 16> resAttrs;
+  mod.getAllResultAttrs(resAttrs);
+
+  // Iterate through the results and append to one of the two below lists. The
+  // first for non-ESI-ports. The second, ports which have been re-located to an
+  // operand.
+  SmallVector<Type, 8> newResultTypes;
+  SmallVector<MutableDictionaryAttr, 4> newResultAttrs;
+  for (size_t resNum = 0, numRes = funcType.getNumResults(); resNum < numRes;
+       ++resNum) {
+    Type resTy = funcType.getResult(resNum);
+    auto chanTy = resTy.dyn_cast<ChannelPort>();
+    if (!chanTy) {
+      newResultTypes.push_back(resTy);
+      newResultAttrs.push_back(resAttrs[resNum]);
+      continue;
+    }
+
+    // When we find one, construct an interface, and add the 'sink' modport to
+    // the type list.
+    InterfaceOp iface = B->getOrConstructInterface(chanTy);
+    ModportType sourcePort = iface.getModportType(B->source);
+    newArgTypes.push_back(sourcePort);
+    argAttrs.push_back(resAttrs[resNum]);
+    updated = true;
+  }
+
+  if (!updated)
+    return;
+
+  // Set the new types.
+  auto newFuncType = FunctionType::get(newArgTypes, newResultTypes, ctxt);
+  mod.setType(newFuncType);
+  mod.setAllArgAttrs(argAttrs);
+  mod.setAllResultAttrs(newResultAttrs);
+
+  // Find all instances and update them.
+  Operation *scope = SymbolTable::getNearestSymbolTable(mod);
+  llvm::Optional<SymbolTable::UseRange> instances =
+      SymbolTable::getSymbolUses(mod, scope);
+  if (!instances)
+    return;
+  for (auto symUse : *instances) {
+    auto instOp = dyn_cast<InstanceOp>(symUse.getUser());
+    if (instOp)
+      updateInstance(mod, instOp);
+  }
+}
+
+/// Update an instance of an updated module by adding `esi.(un)wrap.iface`
+/// around the instance. Create a new instance at the end from the lists built
+/// up before.
+void ESIPortsPass::updateInstance(RTLExternModuleOp mod, InstanceOp inst) {
+  using namespace circt::sv;
+  OpBuilder instBuilder(inst);
+  FunctionType funcTy = mod.getType();
+  auto loc = inst.getLoc();
```

Comment:
Seems like this and other passes are generally using the same Location for large swaths of generated ops. Does it make sense for ESIRTLBuilder to inherit from ImplicitLocOpBuilder?

https://github.com/llvm/circt/blob/3f97ea344d673c2f8c485c28427551d77d6afd12/include/circt/Support/ImplicitLocOpBuilder.h#L14-L17

Reply:
Or at least use that here for `instBuilder`?

Reply:
Yep. ImplicitLocBuilder really should be in MLIR proper, yes?

Reply:
Yep agreed, it was brought up and sounded like it would be appreciated: https://llvm.discourse.group/t/rfc-add-implicitlocopbuilder-to-builders-h/2166

Reply:
https://reviews.llvm.org/D93717

Reply:
Thanks!